File: backend-legacy-python/app/api/admin/admins.py
"""REST API v1: Admin user management (RBAC). Admin-only endpoints."""

import logging

from fastapi import APIRouter, Depends, Request
from fastapi.responses import JSONResponse
from sqlalchemy import text

from app.database.db import async_session
from app.auth.rbac import require_admin, hash_password, ROLE_LEVELS

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/admins", tags=["admins"])

VALID_ROLES = list(ROLE_LEVELS.keys())


@router.get("")
async def list_admins(admin: dict = Depends(require_admin)):
    """List all admin users."""
    try:
        async with async_session() as session:
            result = await session.execute(
                text("SELECT id, username, role, is_active, last_login, created_at FROM admin_users ORDER BY id")
            )
            admins = []
            for row in result.fetchall():
                admins.append({
                    "id": row[0],
                    "username": row[1],
                    "role": row[2],
                    "is_active": row[3],
                    "last_login": row[4].isoformat() if row[4] else None,
                    "created_at": row[5].isoformat() if row[5] else None,
                })
            return admins
    except Exception as e:
        logger.exception("List admins failed: %s", e)
        return JSONResponse({"error": "Internal server error"}, status_code=500)


@router.post("")
async def create_admin_user(request: Request, admin: dict = Depends(require_admin)):
    """Create a new admin user. Body: {username, password, role}."""
    try:
        body = await request.json()
        username = body.get("username", "").strip()
        password = body.get("password", "")
        role = body.get("role", "viewer")

        if not username or not password:
            return JSONResponse({"error": "username and password required"}, status_code=400)
        if len(username) > 64:
            return JSONResponse({"error": "username too long (max 64)"}, status_code=400)
        if len(password) < 6:
            return JSONResponse({"error": "password too short (min 6)"}, status_code=400)
        if role not in VALID_ROLES:
            return JSONResponse({"error": f"invalid role, must be one of: {VALID_ROLES}"}, status_code=400)

        pw_hash = hash_password(password)

        async with async_session() as session:
            existing = await session.execute(
                text("SELECT id FROM admin_users WHERE username = :u"), {"u": username}
            )
            if existing.fetchone():
                return JSONResponse({"error": "username already exists"}, status_code=409)

            await session.execute(
                text("INSERT INTO admin_users (username, password_hash, role) VALUES (:u, :p, :r)"),
                {"u": username, "p": pw_hash, "r": role},
            )
            await session.commit()

            result = await session.execute(
                text("SELECT id, username, role, is_active, created_at FROM admin_users WHERE username = :u"),
                {"u": username},
            )
            row = result.fetchone()

        logger.info("Admin %s created user %s (role=%s)", admin["username"], username, role)
        return JSONResponse({
            "ok": True,
            "admin": {
                "id": row[0], "username": row[1], "role": row[2],
                "is_active": row[3], "created_at": row[4].isoformat() if row[4] else None,
            },
        }, status_code=201)

    except Exception as e:
        logger.exception("Create admin failed: %s", e)
        return JSONResponse({"error": "Internal server error"}, status_code=500)


@router.patch("/{admin_id}")
async def update_admin_user(admin_id: int, request: Request, admin: dict = Depends(require_admin)):
    """Update admin user. Body: optional {role, is_active, password}."""
    try:
        body = await request.json()

        updates = []
        params: dict = {"id": admin_id}

        if "role" in body:
            if body["role"] not in VALID_ROLES:
                return JSONResponse({"error": "invalid role"}, status_code=400)
            updates.append("role = :role")
            params["role"] = body["role"]

        if "is_active" in body:
            updates.append("is_active = :active")
            params["active"] = bool(body["is_active"])

        if "password" in body:
            if len(body["password"]) < 8:
                return JSONResponse({"error": "password too short (min 8)"}, status_code=400)
            if len(body["password"]) > 128:
                return JSONResponse({"error": "password too long (max 128)"}, status_code=400)
            updates.append("password_hash = :pw")
            params["pw"] = hash_password(body["password"])

        if not updates:
            return JSONResponse({"error": "no fields to update"}, status_code=400)

        # Prevent self-demotion
        if admin.get("user_id") == admin_id and "role" in body and body["role"] != "admin":
            return JSONResponse({"error": "cannot demote yourself"}, status_code=400)

        # Prevent self-deactivation
        if admin.get("user_id") == admin_id and "is_active" in body and not body["is_active"]:
            return JSONResponse({"error": "cannot deactivate yourself"}, status_code=400)

        async with async_session() as session:
            sql = f"UPDATE admin_users SET {', '.join(updates)} WHERE id = :id"
            result = await session.execute(text(sql), params)
            if result.rowcount == 0:
                return JSONResponse({"error": "admin not found"}, status_code=404)
            await session.commit()

        logger.info("Admin %s updated admin_id=%d: %s", admin["username"], admin_id, list(body.keys()))
        return {"ok": True}

    except Exception as e:
        logger.exception("Update admin failed: %s", e)
        return JSONResponse({"error": "Internal server error"}, status_code=500)


@router.delete("/{admin_id}")
async def delete_admin_user(admin_id: int, admin: dict = Depends(require_admin)):
    """Delete admin user. Cannot delete yourself."""
    try:
        if admin.get("user_id") == admin_id:
            return JSONResponse({"error": "cannot delete yourself"}, status_code=400)

        async with async_session() as session:
            result = await session.execute(
                text("DELETE FROM admin_users WHERE id = :id"), {"id": admin_id}
            )
            if result.rowcount == 0:
                return JSONResponse({"error": "admin not found"}, status_code=404)
            await session.commit()

        logger.info("Admin %s deleted admin_id=%d", admin["username"], admin_id)
        return {"ok": True}

    except Exception as e:
        logger.exception("Delete admin failed: %s", e)
        return JSONResponse({"error": "Internal server error"}, status_code=500)

File: backend/crates/chameleon-admin/src/admin/admins.rs
//! Admin user CRUD endpoints.
//! GET /admins, POST /admins, DELETE /admins/{id}

use axum::{extract::{Path, State}, routing::{delete, get, post}, Json, Router};
use serde::Deserialize;

use chameleon_auth::{password, rbac::Role, RequireAdmin};
use chameleon_db::queries::admin as admin_q;
use chameleon_core::{ChameleonCore, error::{ApiError, ApiResult}};

pub fn router() -> Router<ChameleonCore> {
    Router::new()
        .route("/admins", get(list_admins).post(create_admin))
        .route("/admins/{id}", delete(delete_admin))
}

async fn list_admins(
    State(state): State<ChameleonCore>,
    _admin: RequireAdmin,
) -> ApiResult<Json<Vec<serde_json::Value>>> {
    let admins = admin_q::list_admins(&state.db).await.map_err(|e| ApiError::Internal(e))?;
    let result: Vec<serde_json::Value> = admins.into_iter().map(|a| {
        serde_json::json!({
            "id": a.id, "username": a.username, "role": a.role,
            "is_active": a.is_active,
            "last_login": a.last_login.map(|t| t.format("%Y-%m-%dT%H:%M:%S").to_string()),
            "created_at": a.created_at.map(|t| t.format("%Y-%m-%dT%H:%M:%S").to_string()),
        })
    }).collect();
    Ok(Json(result))
}

#[derive(Deserialize)]
struct CreateAdminBody {
    username: String,
    password: String,
    role: Option<String>,
}

async fn create_admin(
    State(state): State<ChameleonCore>,
    admin: RequireAdmin,
    Json(body): Json<CreateAdminBody>,
) -> ApiResult<Json<serde_json::Value>> {
    let username = body.username.trim();
    if username.is_empty() || body.password.is_empty() {
        return Err(ApiError::BadRequest("username and password required".into()));
    }
    if username.len() > 64 {
        return Err(ApiError::BadRequest("username too long (max 64)".into()));
    }
    if body.password.len() < 8 {
        return Err(ApiError::BadRequest("password too short (min 8)".into()));
    }
    if body.password.len() > 128 {
        return Err(ApiError::BadRequest("password too long (max 128)".into()));
    }
    let role = body.role.as_deref().unwrap_or("viewer");
    if Role::from_str(role).is_none() {
        return Err(ApiError::BadRequest("invalid role".into()));
    }

    // Check uniqueness
    if admin_q::find_admin_by_username(&state.db, username).await.map_err(|e| ApiError::Internal(e))?.is_some() {
        return Err(ApiError::Conflict("username already exists".into()));
    }

    let pw_hash = password::hash_password(&body.password)
        .map_err(|e| ApiError::Internal(anyhow::anyhow!("hash error: {e}")))?;

    let new_admin = admin_q::create_admin(&state.db, username, &pw_hash, role)
        .await.map_err(|e| ApiError::Internal(e))?;

    tracing::info!(creator = %admin.0.username, new_user = username, role, "Admin created");

    Ok(Json(serde_json::json!({
        "ok": true,
        "admin": {"id": new_admin.id, "username": new_admin.username, "role": new_admin.role},
    })))
}

async fn delete_admin(
    State(state): State<ChameleonCore>,
    admin: RequireAdmin,
    Path(admin_id): Path<i32>,
) -> ApiResult<Json<serde_json::Value>> {
    if admin.0.user_id == admin_id {
        return Err(ApiError::BadRequest("cannot delete yourself".into()));
    }

    let deleted = admin_q::delete_admin(&state.db, admin_id)
        .await.map_err(|e| ApiError::Internal(e))?;

    if !deleted {
        return Err(ApiError::NotFound("admin not found".into()));
    }

    tracing::info!(admin = %admin.0.username, deleted_id = admin_id, "Admin deleted");
    Ok(Json(serde_json::json!({"ok": true})))
}
